File: kotlin/src/main/kotlin/394.DecodeString.kt
import org.junit.Assert.assertEquals
import org.junit.Test
import java.lang.StringBuilder
import java.util.*

// https://leetcode.com/problems/decode-string


class DecodeString {
  fun decodeString(s: String): String {
    var cur = StringBuilder()
    var num = StringBuilder()
    val digits = "0123456789".toSet()
    val stack = Stack<Pair<StringBuilder, StringBuilder>>()
    for(ch in s)
      when(ch){
        '[' -> {
          stack.push(cur to num)
          cur = StringBuilder()
          num = StringBuilder()
        }
        ']' -> {
          val sub = cur
          val last = stack.pop()
          cur = last.first
          cur.append(sub.repeat(last.second.toString().toInt()))
          num.setLength(0)
        }
        in digits -> num.append(ch)
        else -> cur.append(ch)
      }
    return cur.toString()
  }

  @Test
  fun test1(){
    assertEquals("aaabcbc", decodeString("3[a]2[bc]"))
  }

  @Test
  fun test2(){
    assertEquals("accaccacc", decodeString( "3[a2[c]]"))
  }

  @Test
  fun test3(){
    assertEquals("abcabccdcdcdef", decodeString("2[abc]3[cd]ef"))
  }


}

File: python/394_DecodeString.py
from unittest import TestCase
# https://leetcode.com/problems/decode-string


class DecodeString(TestCase):
    def decodeString(self, s: 'str') -> 'str':
        n = len(s)

        def dec(i):
            pre, res = None, []
            while i < n and s[i] != ']':
                ch = s[i]
                if ch == '[':
                    num = int(s[pre:i])
                    sub, i = dec(i+1)
                    res += num*sub
                    pre = None
                elif not ch.isdigit():
                    res.append(ch)
                elif pre is None:
                    pre = i
                i += 1
            return res, i
        return "".join(dec(0)[0])

    # # iter
    # def decodeString(self, s: 'str') -> 'str':
    #     st, parts, digits = [], [], []
    #     for c in s:
    #         if c.isdigit():
    #             digits.append(c)
    #         elif c == '[':
    #             k = int("".join(digits))
    #             digits = []
    #             st.append(parts)
    #             st.append(k)
    #             parts = []
    #         elif c == ']':
    #             str = "".join(parts)
    #             k = st.pop()
    #             parts = st.pop()
    #             parts.append(k*str)
    #         else:
    #             parts.append(c)
    #     return "".join(parts)


    def test1(self):
        self.assertEqual("aaabcbc", self.decodeString("3[a]2[bc]"))

    def test2(self):
        self.assertEqual("accaccacc", self.decodeString("3[a2[c]]"))

    def test3(self):
        self.assertEqual("abcabccdcdcdef", self.decodeString("2[abc]3[cd]ef"))

